CCS with minimize_roundtrips performs incremental merges of each SearchResponse

quux00 opened the pull request:
To help address the issue of slow-to-respond clusters in a cross-cluster search,
async-search based CCS with minimize_roundtrips=true performs incremental
merges of each SearchResponse as they come in from each cluster (including
the local cluster).

This means, any time the user calls GET _async_search/:id, they will now get
search hits and/or aggregation results from any clusters that have finished so far.

The `is_running` field in the async-search response should be used to determine
whether at least one cluster has still not reported back.

quux00 committed:
CCS with minimize_roundtrips performs incremental merges of each SearchResponse

To help address the issue of slow-to-respond clusters in a cross-cluster search,
async-search based CCS with minimize_roundtrips=true performs incremental
merges of each SearchResponse as they come in from each cluster (including
the local cluster).

This means, any time the user calls GET _async_search/:id, they will now get
search hits and/or aggregation results from any clusters that have finished so far.

The `is_running` field in the async-search response should be used to determine
whether at least one cluster has still not reported back.

quux00 committed:
calling notifyCcsReduce on SearchProgressListener with incremental merged results

quux00 committed:
Updated TransportSearchActionTests to test for incrementalResponses

quux00 committed:
Added incRef and decRef to SearchResponses collected by the SearchProgressListener

in the TransportSearchAction.testCCSRemoteReduce test.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/103134.yaml

javanna commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -775,9 +780,13 @@ private static ActionListener<SearchResponse> createCCSListener(
-                // TODO: in CCS fail fast ticket we may need to fail the query if the cluster gets marked as FAILED
                 ccsClusterInfoUpdate(searchResponse, clusters, clusterAlias, skipUnavailable);
                 searchResponseMerger.add(searchResponse);
+                if (progressListener != null && progressListener != SearchProgressListener.NOOP) {
```
can you remind me when the progress listener is actually null?

quux00 replied:
I think there were tests that were doing it, but I have the null check and cleaned up the tests. So far it looks good so I have that in the next push.

javanna replied:
great!

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -107,6 +107,19 @@ synchronized void updatePartialResponse(
+        failIfFrozen();
+
+        this.responseHeaders = threadContext.getResponseHeaders();
+        this.finalResponse = response;
```
I think that this ends up not playing well with partial results coming from the local cluster. Or do we just accept that when minimizing roundtrips with async search we lose that feature?

Also, semantic wise, it may get confusing to set this partial response to the finalResponse instance member. 

Lastly, I was trying to think of the implications/risks: similar to partially aggregated results, they are kept in memory. But search response includes hits as well, hence this may take quite a bit of memory that can't be freed until the search is completed. @original-brownbear do you see a way around this?

javanna replied:
On the semantic side of it, I think that get async status will return isRunning false and isPartial false if you set the partial response to the final response. It seems easy enough to fix, but maybe an indication that this may be misleading and error prone.

On the memory usage, I believe this could be a serious issue especially targeting remote clusters that may never respond (which is exactly the reason for this change). While you are able to retrieve partial results via get async search API, those will be kept in memory on the coordinating node until the search is completed (as we don't store intermediate results - see  #55464) . That's a bit of a risk as 1. the response can be big (with minimize roundtrips we get from+size hits from each cluster) 2. it may take a while, minutes, until all remote clusters have come back, and it may even happen that they don't.  I am thinking this will need some additional testing and thoughts around cleaning memory up.

javanna replied:
More thoughts about memory usage: the partial search response is already kept in memory via SearchResponseMerger for later merge, so perhaps this is an already existing problem. We are just adding a new reference to it in MutableSearchResponse and making it accessible from get async search API, so we may be ok. Hopefully we are not introducing a new problem. What do you both think?

javanna replied:
one more thought: I think that we need to reduce the response because it contains from+size hits, while the user requested size . That suggests we need a test with from higher than `0`.

quux00 replied:
A lot to respond to here. We need slack threads for this one :-)

> I think that this ends up not playing well with partial results coming from the local cluster. Or do we just accept that when minimizing roundtrips with async search we lose that feature?

We partially lose that feature with MRT=true full merges.

With MRT=true, the local cluster also gets a CCSActionListener, so it also calls `onCcsMinimizeRoundtripsReduce(SearchResponse)` when it has finished, merging its full results with whatever other clusters have responded. 

So a typical flow is (from some logging I added to one of the IT tests):

```
Creating CCSListener for cluster [cluster_a]
Creating CCSListener for cluster []
onListShards called: shards size: 9
onPartialReduce called: shards size: 5
onFinalReduce called: shards size: 9
TSA notifyCcsMinimizeRoundtripsReduce cluster []
TSA notifyCcsMinimizeRoundtripsReduce cluster [cluster_a]
TSA createFinalResponse cluster [cluster_a]
```

So we will still have the partial aggs results from the local cluster, but only as long as no other remote cluster has returned its results. Once any full search has finished, then we ignore the partial aggs from the local cluster.

Since in `MutableSearchResponse.toAsyncSearchResponse` we create a new SearchResponse from just InternalAggs (if finalResponse is not set), we could have additional logic to merge in that SearchResponse, but that is a fair bit of work since: (1) MutableSearchResponse does not have access to a SearchResponseMerger and (2) we would have to track whether the local cluster has finished. (I guess you could tell that from when onFinalReduce is called.)

So if we wanted to do that I think it's feasible but we'd probably need a second SearchResponseMerger in the MutableSearchResponse, rather than have the MutableSearchResponse have access to the on in the TransportSearchAction.
- the Mutable

quux00 replied:
> On the semantic side of it, I think that get async status will return isRunning false and isPartial false if you set the partial response to the final response. 

Good catch. I can adjust that by testing against the frozen flag. I'll try that now.

quux00 replied:
> More thoughts about memory usage: the partial search response is already kept in memory via SearchResponseMerger for later merge, so perhaps this is an already existing problem. We are just adding a new reference to it in MutableSearchResponse and making it accessible from get async search API, so we may be ok. Hopefully we are not introducing a new problem. What do you both think?

Yes, SearchResponseMerger already holds all the individual cluster SearchResponses in memory. This PR creates one additional merged SearchResponse from each of those and holds that one in memory in the MutableSeachResponse, overwriting the reference to the last one. I believe @original-brownbear will be added decRef logic to MutableSearchResponse in a later PR, so I haven't done that hear.

So this PR does create one additional (possibly large) SearchResponse that is held in memory. Three options here:
1) because of Armin's shard array work on SearchResponse, this may turn out to not be much more RAM usage (?)
2) if it is more RAM usage, then we could change SearchResponseMerger to merge as each SearchResponse comes in and only hold a reference to that intermediate SearchResponse. That would align with the [original implementation](https://github.com/elastic/elasticsearch/pull/102389/files#diff-9b3d13614ee958e69c6088826f4d14821128164ae6489f2f6c917896c8af5f3dR149) I did.
3) start writing asyc-search results to disk as you outlined here: https://github.com/elastic/elasticsearch/issues/55464

javanna replied:
> We could rename it to latestFullResponse or mergedResponse.

I think that's a good approach. Maybe review all the usages to make sure that there is nothing else we are missing, besides the get async status flags.

> Since in MutableSearchResponse.toAsyncSearchResponse we create a new SearchResponse from just InternalAggs (if finalResponse is not set), we could have additional logic to merge in that SearchResponse, but that is a fair bit of work since: (1) MutableSearchResponse does not have access to a SearchResponseMerger and (2) we would have to track whether the local cluster has finished. (I guess you could tell that from when onFinalReduce is called.)

I can see that. There may be different solutions to this. I think it would be nice to look into it and assess how much effort it takes. It is a bit weird that you get partial results from the local cluster, and those disappear once you get results from one remote cluster. ideally we would merge the two, and you could create a new merger on the fly as well with the two responses you have? happy to brainstorm around this.

quux00 replied:
> one more thought: I think that we need to reduce the response because it contains from+size hits, while the user requested size . That suggests we need a test with from higher than 0.

I don't think I understand what you are saying or proposing. A few responses (which may not be on the mark):

1. do you mean do some additional modification in the MutableSearchResponse rather than in the SearchResponseMerger?
2. there are already tests for SearchResponseMerger that test mergers with different from/size settings: [example](https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java#L681). Does that cover what you are suggesting?

javanna replied:
I got confused around the response we are exposing. We exposed what getMergedResponse returns, hence that is properly reduced. The input search responses have from+size results, but the merged response has only size results.

quux00 replied:
>> Since in MutableSearchResponse.toAsyncSearchResponse we create a new SearchResponse from just InternalAggs (if finalResponse is not set), we could have additional logic to merge in that SearchResponse, but that is a fair bit of work since: (1) MutableSearchResponse does not have access to a SearchResponseMerger and (2) we would have to track whether the local cluster has finished. (I guess you could tell that from when onFinalReduce is called.)

> I can see that. There may be different solutions to this. I think it would be nice to look into it and assess how much effort it takes. It is a bit weird that you get partial results from the local cluster, and those disappear once you get results from one remote cluster. ideally we would merge the two, and you could create a new merger on the fly as well with the two responses you have? happy to brainstorm around this.

I made a  first attempt at this (not pushed to this PR) was to have the SearchResponseMerger implement SearchProgressListener, so that it could get the partial results of the local cluster. But that doesn't work with the TransportSearchAction design since the SearchProgressListener is not passed in directly down the stack to the local search code. But instead, the SearchProgressListener is embedded in the task. And the AsyncSearchTask is separated by quite a few layers from the where the SearchResponseMerger is created (in fact TransportSearchAction never explicitly knows whether the search is async or not).

An alternative design could be to have the AsyncSearchTask have a reference to the same SearchResponseMerger and have it do the incremental merges, rather than the CCS ActionListener. But that would mean that sync-search and async-search for MRT=true work very differently.

So I think I could benefit from a brainstorming, design session if we want to purse this option.

The simplest solutions are:

1. Create a second SearchResponseMerger that is part of the AsyncSearchTask (or MutableSearchResponse) and can do a merger of the partial local results with whatever full SearchResponses have come in so far incrementally. This I think would be reasonable straightforward, but is fairly clunky.

2. For MRT=true, stop doing partial agg reductions for local searches entirely. One could make the argument that with MRT=true, we should treat all clusters the same - we wait until the entire result set arrives from each cluster and then do an incremental merge of those full restults. Then we don't have partial results from the local cluster showing up and then temporarily disappearing. This change would also have no negative impact on Kibana users, since Kibana does not show incremental results yet anyway.

I think my vote is actually for the last option above. I think that is cleanest - both code wise as well as in terms of explaining functionality to end users.

quux00 replied:
I've updated the `finalResponse` field in MutableSearchResponse to be called `latestFullResponse` in [this commit](https://github.com/elastic/elasticsearch/pull/103134/commits/c7cd36293fa067b2d42aa59eb6c0261b88829e1a).

And I updated MutableSearchResponse in [this last commit](https://github.com/elastic/elasticsearch/pull/103134/commits/901c2f13f043c66009c6c58fc804e6317fe7c137) to ignore partial aggs results from the local cluster when CCS MRT=true in order to test out my suggested approach above. It passes all the automated tests I tried it against and I spent about 30 minutes manually testing local-only, CCS MRT=false and CCS MRT=true with ~15 shards per index and it looks like it behaves correctly.

javanna reviewed (commented):
Left some comments, this looks sensible to me.

quux00 committed:
PR feedback (partial)

quux00 commented on server/src/test/java/org/elasticsearch/action/search/TransportSearchActionTests.java:
```diff
@@ -592,6 +594,18 @@ public void testCCSRemoteReduce() throws Exception {
                     }),
                     latch
                 );
+                List<SearchResponse> incrementalResponses = new ArrayList<>();
```
Thanks for catching my concurrency bug - fixed.

I can split those tests out in this PR. I'll do that in the next push.

quux00 committed:
PR Feedback

Renamed onCcsReduce to onCcsMinimizeRoundtripsReduce
Fixed bug in AsyncStatusResponse that it now checks the frozen flag to determine
is_running and is_partial values.
Split TransportSearchActionTests.testCCSRemoteReduce into multiple test methods.

quux00 committed:
Minor test adjustment - changing from/size on one of the TransportSearchActionTests.testCCSRemoteReduce subtests

quux00 committed:
In MutableSearchResponse, renamed finalResponse to latestFullResponse

And updated javadoc in various places to reflect to the new model for CCS MRT=true

quux00 committed:
For MRT=true, stop doing partial agg reductions for local searches entirely.

With MRT=true, we should treat all clusters the same - we wait until the entire
result set arrives from each cluster and then do an incremental merge of those
full restults. Then we don't have partial results from the local cluster showing
up and then temporarily disappearing. This change would also have no negative impact
on Kibana users, since Kibana does not show incremental results yet anyway.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -107,6 +121,20 @@ synchronized void updatePartialResponse(
+     * @param response latest full response with SearchResponses from all
+     *                 clusters that have reported back
+     */
+    synchronized void updatePartialResponse(SearchResponse response) {
```
Should we include CCSMinimizeRoundtrips in this method's name too?

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -232,19 +260,19 @@ synchronized AsyncStatusResponse toStatusResponse(String asyncExecutionId, long
-                false,
-                false,
+                frozen == false,
+                frozen == false,
```
I still get confused about this: should we look at isPartial for the isPartial flag?

quux00 replied:
Yes, you are correct. Good catch. Fixed in next push.

javanna commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -775,9 +780,13 @@ private static ActionListener<SearchResponse> createCCSListener(
+                if (progressListener != SearchProgressListener.NOOP) {
+                    // do an incremental merge of all SearchResponses received so far for async-search with CCS MRT=true
+                    // (sync search uses SearchProgressListener.NOOP, so this path will not execute)
+                    progressListener.notifyCcsMinimizeRoundtripsReduce(searchResponseMerger.getMergedResponse(clusters));
```
Ok I finally understood what you meant in another comment: we do hold an additional search response in memory, because search response merger still does what it's supposed to do and holds all the search responses from all remote clusters until they all come back. In addition, we do a partial merge to update the async search result.

An alternative if we want to save memory is to delay the merging until results are retrieved, similar to how we deal with partial aggs results currently. It does require some changes to your current PR. We could either to it every time, or cache the result after the first get async search call (like we do for aggs if I am not mistaken).

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -95,6 +106,9 @@ synchronized void updatePartialResponse(
         failIfFrozen();
+        if (clusters.isCcsMinimizeRoundtrips()) {
+            return;
+        }
```
This makes sense if we decide that we don't want to expose partial aggs results, much better than them disappearing once a remote cluster returned its response. I have a counter proposal: could we instead look into creating a SearchResponseMerger, on the fly, and add the partial SearchResponse together with the partial aggs and expose the result of the merge? That should work or am I missing something?

```
InternalAggregations reducedAggs = reducedAggsSource.get();
reducedAggsSource = () -> reducedAggs;
SearchResponse searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
SearchResponseMerger merger = new SearchResponseMerger();
merger.add(searchResponse);
merger.add(partialSearchResponse);
SearchResponse newSearchResponse = merger.getMergedResponse();
```

It feels like this should fit with the other comment above on delaying the merging to once results are requested.

javanna reviewed (commented):
I left some more comments.

quux00 committed:
Setting is_partial state in MutableSearchResponse based on isPartial flag

quux00 committed:
This commit reverses the previous logic of pushing merged changes to the AsyncSearchTask/MutableSearchResponse.

Now the MutableSearchResponse can pull incremental merges on-demand,
so the merged incremental response no longer needs to be retained in memory
(it is a local variable now).

In addition, we retain the ability to use the partial aggs from the local
cluster until it completes a full reduction as part of the incremental results.

By doing this inversion of logic, the SearchProgressListener no longer requires
a new `onCCSMinimizeRoundtripsReduce` callback. Instead, the key to this design
is to allow the AsyncSearchTask/MutableSearchResponse to have a reference
to the same SearchResponseMerger that the TransportSearchAction creates.
To do this, the creation of that SearchResponseMerger is now done in the
SearchTask and the AsyncSearchTask overrides that method to retain a reference
to the SearchResponseMerger and pass it to the MutableSearchResponse.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -197,15 +213,29 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
+                reducedAggsSource = () -> reducedAggs;
+                // TODO: check that this just returns the partial aggs response passed in if SearchResponseMerger
+                // has no other full SearchResponses
+                searchResponse = searchResponseMerger.getMergedResponse(clusters, buildResponse(task.getStartTimeNanos(), reducedAggs));
```
You could hold a reference to ` Supplier<SearchResponse>` instead to that we do the same as we do for aggs: delay reduction until results are requested, hold the reduced results in memory for later retrieval once reduced for the first time.

quux00 replied:
I don't think I follow.

First, this does hold the `reducedAggsSource` as an instance variable just like the current approach in main.

Second, the SearchResponse is not currently cached (in main) - it is always regenerated on demand and assigned only to a local variable. That is the same in my new impl.

Third, I don't think we want to cache a SearchResponse - that's a key reason for the reorg of the code in my last commit - to not hold more RAM in an instance var.

Fourth, I don't think we want to cache SearchResponses because the call to SearchResponseMerger may generate a different response if more clusters have reported in. We need to always generate a new response until the final response is written to disk/index.

I have a feeling that none of my above comments are addressing what you suggest though, but I'm unclear on what value a `Supplier<SearchResponse>` brings here. We already effectively have that in a SearchResponseMerger reference. Why does another layer of indirection help?

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -92,7 +104,8 @@ synchronized void updatePartialResponse(
         Supplier<InternalAggregations> reducedAggs,
-        int reducePhase
+        int reducePhase,
+        boolean finalReduce /// MP TODO: document me
```
This looks odd: updatePartialResponse suggests partial reduction, and it taking a flag `finalReduce` suggests otherwise.

quux00 replied:
That is actually already the case, as AsyncSearchTask calls this method from both
[onPartialReduce](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java#L470) and [onFinalReduce](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java#L484).

It's just that the MutableSearchResponse right now doesn't need to know whether it is the partial or final reduce of the aggs and it doesn't matter for the existing merge logic.

But for the new incremental merge logic for MRT=true, we know need to know in order to determine whether or not to use the partial reduce of local-search aggs. Once the final reduce is true, we no longer want to use the reduction passed in here. We instead want to use the full SearchResponse that is added to the SearchResponseMerge by the CCSActionListener (otherwise we would have double counting/merging of the local results).

If you really don't want this `finalReduce` boolean param here, then we would need to create a new `updateFinalResponse` in MutableSearchResponse that AsyncSearchTask.onFinalReduce would call, but it would do the same thing as `updatePartialResponse` other than setting this new `finalReduce` flag to true.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -197,15 +213,29 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
+                InternalAggregations reducedAggs = reducedAggsSource.get();
+                reducedAggsSource = () -> reducedAggs;
+                searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            } else if (finalReduce == false) {
```
Using this flag looks odd, mostly because this method should always ever expose finally reduced results, for user consumption.

quux00 replied:
> this method should always ever expose finally reduced results, for user consumption.

No, that's not the case currently. Async search will show partially reduced results. For example, in the example below only 9 of 15 shards from the local cluster have reported back and we show those results (this example is from a test of code on `main` not this branch).

<details>
  <summary><b>Partial results of local cluster only</b> (toggle to view)</summary>

```
{
  "id": "FjNiUEszYnQ4U2thR1M1VjNOR0FVMEEaS2JNdEllR3hUVkdNMWh1aUozOF83UTo0MzA=",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1702916562276,
  "expiration_time_in_millis": 1703348562276,
  "response": {
    "took": 11355,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 2,
    "_shards": {
      "total": 15,
      "successful": 9,  // 9/15 shards on the local cluster
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 0,
      "skipped": 0,
      "running": 3,
      "partial": 0,
      "failed": 0,
      "details": {
        "(local)": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        },
        "remote2": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        },
        "remote1": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        }
      }
    },
    "hits": {
      "total": {
        "value": 752,
        "relation": "eq"
      },
      "max_score": null,
      "hits": []
    },
    "aggregations": {
      "indexgroup": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
          {
            "key": "blogs",
            "doc_count": 737  // partial results - there are 1400 docs total
          },
          {
            "key": "web_traffic",
            "doc_count": 15
          }
        ]
      }
    }
  }
}

```
</details>

For CCS MRT=true, the final local cluster results are reported in two ways: via the onPartialReduce and onFinalReduce callbacks to the SearchProgressListener (coming into the `MutableSearchResponse.updatePartialResponse` method) and then the CCSActionListener for the local cluster sends the final full `SearchResponse` to the SearchResponseMerger.

So here we don't want to use the fully reduced results that came into `MutableSearchResponse.updatePartialResponse` because we are going to have them from the SearchResponseMerger list (otherwise those local results would be included twice in the merged response).

The logic here is:

```
if (searchResponseMerger == null) {
    // this is a local-only search or CCS MRT=false, so process the results
    // coming in from onPartialReduce as we do currently in 'main'
} else if (finalReduce == false) {
    // we are doing CCS MRT=true and the local cluster has reported back
    // only partial results (subset of shards), so do a SearchResponseMerger-
    // based merge of any full results that have come in from remote clusters 
    // and the partial results of the local cluster
} else {
    // we are doing CCS MRT=true and the local cluster has reported back
    // full results (via both (1) updatePartialResults and (2) the CCSActionListener
    // passing the full local-cluster SearchResponse to the SearchResponseMerger)
    // so create a merged result based on full responses held in SearchResponseMerger
    // received so far (may still be running, may be completed)
}
```

javanna replied:
> No, that's not the case currently. Async search will show partially reduced results.

Just to clarify what I meant: async search with show incremental results, but only finally reduced results are ever returned to the user via get async search.

Perhaps the confusion on my end is around naming. I understand now the difference between final reduction done locally, and updateFinalResponse which is called when all clusters are done, in the context of CCS. Perhaps we could rename the flag to localFinalReduce? One thing I am a bit fuzzy on though is whether we are ok ignoring partial aggs results if finalReduce is true. Doesn't the supplier get assigned anyways in updatePartialResponse ?

quux00 replied:
> Perhaps we could rename the flag to localFinalReduce?
Good idea. Fixed in next push.

quux00 replied:
> One thing I am a bit fuzzy on though is whether we are ok ignoring partial aggs results if finalReduce is true. 

I believe so. Once `localFinalReduce` is true (and CCS MRT=true), we no longer need to use the data held in `reducedAggsSource` because the final SearchResponse from the local cluster has been sent to the SearchResponseMerger (in TransportSearchAction, the local cluster [has its own CCSActionListener](https://github.com/elastic/elasticsearch/blob/0570b0baaa509436555c5711ba1e36d228e4da93/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L615)), so SearchResponseMerger.getMergedResponse will return the local results (hits and aggs). We no longer need to provide the local search response to it.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -437,10 +452,15 @@ protected void onListShards(
+                clusters,
+                threadPool.getThreadContext()
             );
+            mutableSearchResponse.setSearchResponseMerger(searchResponseMerger); // MP TODO: should we just add this to the ctor?
```
I was also going to leave a similar comment to yours (should we add this to the ctor?), but I left a comment below around using a supplier, that may actually help move the dependency to SearchResponseMerger as an argument of updatePartialResponse, perhaps. I have not thought about it a lot but it may make sense to consider that. An alternative could be a static method like `InternalAggregations#toplevelReduce` that takes all the things to be merged together as an argument.

quux00 replied:
I may not be seeing the full picture of what you are suggesting (see my other comment about my confusion), but a few responses:

> move the dependency to SearchResponseMerger as an argument of updatePartialResponse

If this means to allow the SearchResponseMerger to hold/cache the results of the local cluster results of InternalAggs (or any results that come from the data sent to `MutableSearchResponse. updatePartialResponse`, we need to consider that we don't want to retain that result, because the actual full final result (for CCS MRT=true) comes in via CCSActionListener for the local cluster to the SearchResponseMerger (see my other comments for deeper dive into that).

In addition, I discovered a bug in this commit, where SearchResponseMerger was being set as `null` on the MutableSearchResponse when doing a CCS with remote clusters only (no local shards) because in that case onListShards is called before the call to `AsyncSearchTask.getSearchResponseMergerSupplier`, since the onListShards is called "manually" [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L363).

So in my follow-on commit, I have kept this as a setter, not part of the constructor, since I also need this code in `AsyncSearchTask.getSearchResponseMergerSupplier`:

```java
    @Override
    public Supplier<SearchResponseMerger> getSearchResponseMergerSupplier(
        SearchSourceBuilder source,
        TransportSearchAction.SearchTimeProvider timeProvider,
        AggregationReduceContext.Builder aggReduceContextBuilder
    ) {
        return () -> {
            searchResponseMerger = SearchTask.createSearchResponseMerger(source, timeProvider, aggReduceContextBuilder);
            MutableSearchResponse mutableSearchResponse = searchResponse.get();
            // when a remote-only CCS is done (no local shards), then onListShards is called BEFORE
            // getSearchResponseMergerSupplier, so check if the MutableSearchResponse has been created
            // and if so, set the SearchResponseMerger
            if (mutableSearchResponse != null) {
                mutableSearchResponse.setSearchResponseMerger(searchResponseMerger);
            }
            return searchResponseMerger;
        };
    }

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:
```diff
@@ -110,13 +110,20 @@ int numResponses() {
+    /**
+     * TODO: DOCUMENT ME
+     */
+    public SearchResponse getMergedResponse(Clusters clusters, SearchResponse additionalResponse) {
```
I feel like this new method complicates a bit the design of `SearchResponseMerger`. What are the downsides of creating a new one providing it all the responses + the artifically built one with aggs? As opposed to the existing one, it should be a short lived object? Perhaps we could expose that as a static method?

quux00 replied:
That would work, but seems less desirable to me. The core goal of the new refactoring was to share a single SearchResponseMerger. In addition, this new idea would more-or-less break encapsulation as we now have to pass out all of the inner state of the SearchResponseMerger to another one. I guess I could do it as a copy constructor.

javanna reviewed (commented):
This is a good direction, I left some comments for improvements and ideas.

quux00 committed:
Fixes bug in AsyncSearchTask - need to ensure that SearchResponseMerger is set when no local shards being searched.

Removed System.err.println so that build can pass and ensure all tests are passing.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:
```diff
@@ -99,7 +99,7 @@ final class SearchResponseMerger implements Releasable {
      * That may change in the future as it's possible to introduce incremental merges as responses come in if necessary.
      */
-    void add(SearchResponse searchResponse) {
+    public void add(SearchResponse searchResponse) {
```
this can go back to package private I think

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchTask.java:
```diff
@@ -53,4 +57,36 @@ public final SearchProgressListener getProgressListener() {
         return progressListener;
     }
 
+    public Supplier<SearchResponseMerger> getSearchResponseMergerSupplier(
```
this could also be package private instead of public?

quux00 replied:
package private doesn't work because AsyncSearchTask needs to override it and it is in a different package (xpack). But `protected` seems to work, so I'll adjust it to that.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -132,6 +135,25 @@ Listener getSearchProgressActionListener() {
+            if (mutableSearchResponse != null) {
+                mutableSearchResponse.setSearchResponseMerger(searchResponseMerger);
+            }
+            return searchResponseMerger;
```
I find this method a bit hard to reason about. I have a counter proposal: how about we set the `SearchResponseMerger` to the `SearchTask` instead of setting it to the `MutableSearchResponse`. `toAsyncSearchResponse` takes the task as an argument hence the merger could be retrieved from it. This perhaps removes the need for a supplier, and for moving where the merger is created? Should we create it as we used to and set it to the task instead?

quux00 replied:
Yes! Excellent idea. Will be in the next push.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -197,15 +212,26 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
+                // has no other full SearchResponses
+                searchResponse = searchResponseMerger.getMergedResponse(clusters, buildResponse(task.getStartTimeNanos(), reducedAggs));
+            } else {
+                searchResponse = searchResponseMerger.getMergedResponse(clusters);
```
should we assert that there are no partial aggs results in this scenario?

quux00 replied:
No, there could be partial aggs results. We are just ignoring them on this branch because `localFinalReduce`=`true` (and CCS MRT=true). As described in the answer above, we can now ignore any partial results because the SearchResponseMerger has the (full/final) local cluster results.

javanna replied:
That sounds correct. I think that there could be scenarios where finalReduction has happened but the merger did not receive yet those local final results. If that's correct, do we just accept that we won't include such final results until the merger has been notified, hence there will be a little internal delay with CCS and MRT=true?

quux00 replied:
Yes, I thought about that too. We'll need to decide if it is worth adding some sort additional logic to handle that slight gap.

*Considerations:*

We would have to add in some logic to determine whether or not the CCSActionListener for the local cluster has received the final SearchResponse (and added it to the SearchResponseMerger). There isn't anything in the SearchResponse passed to `SearchResponseMerger.add` that indicates what cluster it came from. So we could add a second `add` method (`addSearchResponseFromLocalCluster`) and then SearchResponseMerger could have a flag that is used by the MutableSearchResponse to determine whether to use the local partial aggs or not.

That would be straightforward, but again we are adding even more special handling for the local cluster in a cross-cluster search. We have one extreme (very clean codewise) where we don't do partial aggs for local cluster when CCS MRT=true -- all clusters are treated the same and the other where we have extra logic and tracking for the local cluster. Given that many/most large CCS customers don't have data on the local querying cluster, I'm in favor of the "cleaner" approach.

So options are:
1. don't do partial aggs for local when MRT=true
2. live with this slight gap in time such that in some cases the local cluster aggs could disappear from results we had in a previous response
3. add in additional logic to ensure that gap does not occur (tracking whether the local cluster has reported in via CCSActionListener/SearchResponseMerger).

javanna replied:
I have been thinking about alternatives: the root of the issue is that we introduced a separate path for minimize roundtrips that take local results from a different place, depending on a flag which is set to make the switch. But the flag says results are available in the merger before they actually are. I think finding a good solution for this is important. 

In short, I am thinking that instead of synchronizing, we should try to address the issue at the root, so that the problem we have to deal with no longer exists, if possible. One way to get there would be to go back to an additional callback in the progress listener, that is specific to getting results from remote clusters (and not the local one, as those come in via the existing callbacks). that can still use the search response merger under the hood to expose incremental results, but it would not rely on the same merger instance that is already created in TransportSearchAction, and it would not use the merger as an additional way to get notified of new response being made available. It would merely use it to merge response. 

I was considering ways to possibly using the merger for remote responses but not the local one, without going back to the additional progress listener callback, but I don't think that is possible.

quux00 replied:
> > One way to get there would be to go back to an additional callback in the progress listener, that is specific to getting results from remote clusters (and not the local one, as those come in via the existing callbacks). 

What is this new callback sending? Is it sending individual `SearchResponse` objects? Is it sending merged SearchResponse objects like I had in an earlier iteration? Or only SearchResponse objects merged across all remote-only clusters?

What is the callback receiver (AsyncSearchTask/MutableSearchResponse) going to do with that callback object? Since it is being pushed when the clusters report back rather than pulled when the user is requesting an AsyncSearchResponse, we will have to cache it. But we discussed in my earlier iteration that we don't want to cache the newly merged SearchResponses to avoid using more memory.

If the callback receiver is receiving individual per-cluster SearchResponse objects, then we are double caching those between the two SearchResponseMergers you propose. At that point I don't see what the point of the TransportSearchAction-based SearchResponseMerger is. It is redundant.

> that can still use the search response merger under the hood to expose incremental results, but it would not rely on the same merger instance that is already created in TransportSearchAction, and it would not use the merger as an additional way to get notified of new response being made available. It would merely use it to merge response.

Same comment as above. I don't see how two SearchResponseMergers is better, only worse. The problem becomes how does the 2nd SearchResponseMerger get access to the finished responses without using additional RAM (caching)?

quux00 replied:
I have a new idea and a new commit that I like (I hope you do too!): https://github.com/elastic/elasticsearch/pull/103134/commits/289955a6e1b3aa59c6a1f676edc7ade0a691f049

I think this is a pretty clean solution. It does require the `SearchResponseMerger` to become aware that it may be passed a SearchResponse from the local cluster that is a temporary partial response. 

The MutableSearchResponse object can ask the SearchResponseMerger: do you have a result from the local cluster? If not, it creates a SearchResponse from the local partial aggs and passes that in. But there's a race condition (that I didn't handle before) where we might have local results twice, so the `SearchResponseMerger.getMergedResponse` method now checks whether the final local SearchResponse has come in and if so, doesn't use the local partial aggs response.

To my thinking, this expansion of awareness is not out of place - the SearchResponseMerger is only used for MRT=true and its job is to merge responses. It now participates in doing partial result merges safely.

javanna replied:
I'd like to better understand the implications around an additional callback in the progress listener. The problem we used to have with memory usage was around caching the finally reduced response (what getMergedResponse returns), but we are no longer doing that. Getting notified of a new search response that came in, that the merger needs to track too, only adds a new reference to an object that is already held in memory? The important bit is that we don't end up adding new instances of SearchResponses to the heap, which I don't think we would do. What am I missing?

The atomic boolean is ok, but I think we are adding complexity to the search response merger and making it become a component that async search relies on, not only for merging responses but to get somehow notified of new results. I don't like that, because it introduces weakness and complexity around needing to monitor results from two different places. I think it would be much cleaner to have all the logic about the progress of async search in one place. Also, given you need to introduce the distinction between local and remote results in the merger, that I don't think belongs there, it feels like remote results should have their own callback in the progress listener, as this would clarify the contract and make things easier to follow.

First though, let's clarify the memory issue, because indeed that is the most important aspect and why we have moved to the current approach.

quux00 replied:
> let's clarify the memory issue,

You are correct that having MutableSearchResponse/AsyncSearchTask hold onto the same SearchResponses as SearchResponseMerger does not increase memory (where a newly merged SearchResponse would). However, ti does require additional ref-counting. It also potentially opens the question from maintainers of why are we using two different SearchResponseMergers which look somewhat redundant.

Taking your feedback, I've pushed two new commits as next attempts.

**Latest code changes**

I reverted the previous commit where SearchResponseMerger held logic to decide whether to use the localPartialAggs or the local SearchResponse.

The [first new commit](https://github.com/elastic/elasticsearch/pull/103134/commits/7dea93a54516de9df09a7f4caf443f698a4cd3e7) adds in a SearchProgressListener callback so that MutableSearchResponse can track accurately whether the SearchResponse for the local cluster has been added to the (single) SearchResponseMerger. In this one, the MutableSearchResponse does not have a second SearchResponseMerger. The problem with this commit is that it reintroduces the race condition between using local-partial-aggs vs. local full results and could end up using both. 

The [second new commit](https://github.com/elastic/elasticsearch/pull/103134/commits/298e533d401214d955aa0328ec643e5b35a50e1e) changes the new SearchProgressListener callback to also include the SearchResponse from each cluster, so those can be collected by AsyncSearchTask/MutableSearchResponse, which will create their own SearchResponseMerger and solve the race condition above without having to add new logic to SearchResponseMerger.

There are several different ways to structure this. I picked one that seemed like the least amount of code changes and has consistency guarantees (e.g., avoids the race condition).

The problems to solve are:

• how do you provide the constructor arguments to SearchResponseMerger, many of which are created in the TransportSearchAction

• who should collect and hold the SearchResponses passed into the new SearchProgressListener - AsyncSearchTask, MutableSearchResponse, or maybe some new PartialResponseHandler class that they delegate to?

• should MutableSearchResponse create a new SearchResponseMerger on the fly or create a single one that is reused? 

If we create a single one that is reused then it has to be updated as each SearchResponse comes into the SearchProgressListener, so maybe it should be the thing that holds the cluster SearchResponses? But that feels pretty redundant to the other SearchResponseMerger held by the TransportSearchAction/CCSActionListener's. And (unless we apply synchronization) it doesn't solve the race condition between having partial local aggs and local full results.

• where should logic for how to create partial results reside? In my latest commit, I've left it in the "partial results branch" of MutableSearchResponse.toAsyncSearchResponse, but perhaps we should create some PartialResponseHandler as I suggested above to handle that.


Also: I can't merge in `main` any longer, as Armin has added ref-counting to this part of the code and I'll have merge conflicts to deal with, so I'd rather put that off until we agree on a final design for this PR.

javanna replied:
I am not too concerned with questions around using two different search response mergers. I am more concerned about using the merger to signal that new responses have come in, which is difficult to reason about because there are then two ways that results may come from.

I am happy with the additional callback you added and the mechanism you built to decide when partial aggs have to be taken into account or not. Perhaps in the future this will change once we expose partial top hits for local results too.

As for the problems that you listed for solving, I don't have strong preferences on either of them, and I'd leave it up to you to polish this and finish it up, unless there are challenges about picking one approach over the other.

Thanks!

javanna reviewed (commented):
I left a couple of small comments and questions, the overall approach looks good to me.

quux00 committed:
PR feedback

Reduced visibility of SearchResponseMerger.add and
Added docs for SearchResponseMerger and new logic flow in MutableSearchResponse.toAsyncSearchResponse.
MutableSearchResponse no longer has SearchResponseMerger set as an instance var but grabs it from
the AsyncSearchTask as needed.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -132,6 +135,22 @@ Listener getSearchProgressActionListener() {
     }
 
+    @Override
+    protected Supplier<SearchResponseMerger> getSearchResponseMergerSupplier(
```
Do we still need a supplier? I find using a supplier to set a local variable slightly hard to read. Would it make sense to move back the creation of the merger where it was, and have a setter and a getter for it on the task? The set and get on SearchTask could be no-op, while AsyncSearchTask tracks it like it already does?

quux00 replied:
Yeah, good idea. I like that. I put the setter/getter on SearchTask. It will be set when CCS MRT=true and false otherwise.

quux00 replied:
Also it wouldn't be a no-op in SearchTask, since synchronous search for MRT=true also uses the SearchResponseMerger. The distinction is between MRT=true and MRT=false (or local-only).

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -437,10 +455,14 @@ protected void onListShards(
+                clusters,
+                threadPool.getThreadContext()
             );
+            searchResponse.compareAndSet(null, mutableSearchResponse);
```
Just to highlight the actual changes we are making, would it make sense to restore this as it was?

quux00 replied:
Yes. Restored in next push.

quux00 replied:
Yes, fixed.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -192,16 +206,37 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
-            reducedAggsSource = () -> reducedAggs;
-            searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            // partial results branch
+            SearchResponseMerger searchResponseMerger = task.getSearchResponseMerger();
```
Looking at this, I am also wondering if instead of retrieving the merger from the task we should rather move the logic to the task and expose some `merge` method instead on the task itself. I don't have a strong opinion though. I am looking for ways to simplify this block and make it more readable, yet moving it does not necessarily help that :)

quux00 replied:
I think putting it in SearchTask doesn't simplify the logic (as you say) and to me adds in unnecessary role/complexity to AsyncSearchTask, as we'd basically have to have the same merge logic but now in a less direct place?

javanna reviewed (commented):
Thanks, this looks good. I left a few more small things. I think it would be good to document this behaviour too as it could be surprising, and explain very clearly that partial top hits may all be dropped once new results come in.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

quux00 committed:
Removed Supplier<SearchResponseMerger>. We now add the SearchResponseMerger to the SearchTask for MRT=true.

quux00 committed:
Improved docs and removed TODO comment since it is completed

quux00 committed:
Removed commented out code

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

quux00 committed:
Reverted last commit where SearchResponseMerger handles race condition.

Added SearchProgressListener callback to notify when a cluster has finished.
Still uses a single "shared" SearchResponseMerger.

quux00 committed:
notifyClusterResponseMinimizeRoundtrips pushes both clusterAlias and searchResponse

The SearchResponses are collected by MutableSearchResponse.
When a user requests an AsyncSearchResponse, if the final response (from onResponse)
has not been received, then it will create a new SearchResponseMerger on the fly
using the Supplier<SearchResponseMerger> in the SearchTask. This is non-null only
for CCS MRT=true.

quux00 committed:
Removed SearchResponseMerger#getMergedResponse(Clusters clusters, SearchResponse additionalResponse)

added that logic to MutableSearchResposne

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

quux00 committed:
Added in ref-counting of SearchResponses coming from the new callback

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchTask.java:
```diff
@@ -53,4 +54,19 @@ public final SearchProgressListener getProgressListener() {
+     * @return the Supplier of {@link SearchResponseMerger} attached to this task. Will be null
+     * for local-only search and cross-cluster searches with minimize_roundtrips=false.
+     */
+    public Supplier<SearchResponseMerger> getSearchResponseMergerSupplier() {
```
This is a bit odd. Not a huge deal, but perhaps we could look at whether it makes sense to provide what SearchTask needs to create the merger independently. COuld also be done as a follow-up.

quux00 replied:
As I said in my other comment, this is one of the tricky bits of the design. The information needed to create a SearchResponseMerger is:

```
        SearchSourceBuilder source,
        SearchTimeProvider timeProvider,
        AggregationReduceContext.Builder aggReduceContextBuilder
```

None of which are available to the SearchTask and 2 of which are created in the TransportSearchAction, so you would have to create setters to apply all this information into the SearchTask. That seems like a code smell to me (law of Demeter violation). The Supplier seems much cleaner - basically a closure you can call on demand to create new ones as needed.

javanna reviewed (approved):
Left a couple of comments and questions. No blockers. LGTM otherwise.

quux00 commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -227,6 +294,41 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
+        return getMergedResponse(merger, null);
+    }
+
+    private synchronized SearchResponse getMergedResponse(SearchResponseMerger merger, SearchResponse localPartialAggsOnly) {
```
Correct. I'll add `synchronized` to the other `getMergedResponse` to avoid confusion.

quux00 commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -244,8 +346,8 @@ synchronized AsyncStatusResponse toStatusResponse(String asyncExecutionId, long
-                false,
-                false,
+                frozen == false,
+                isPartial,
```
Yes, it is an existing bug that I noticed in testing. The toAsyncStatus behavior does not match toAsyncResponse behavior. I can undo this change here and do a follow-on PR. That will also re-open the discussion of the divergence of is_partial between CCS and local-only that still exists.

quux00 committed:
Removed change to AsyncStatusResponse (bug fix for later PR), dec-ref finalization code changes/cleanup

quux00 committed:
Added testPartialAggsMixedWithFullResponses to SearchResponseMergerTests and minor code cleanups before merging

quux00 committed:
Removed TODO

quux00 committed:
Merge remote-tracking branch 'elastic/main' into ccs/mrt-full-merges2
